stumpy/stump_tiles.py

```diff
@@ -0,0 +1,780 @@
+    <https://www.cs.ucr.edu/~eamonn/public/GPU_Matrix_profile_VLDB_30DraftOnly.pdf>`__
+
+    See Section 3.1 and Section 3.3
+
+    The above reference outlines the use of the Pearson correlation via Welford's
+    centered sum-of-products along each diagonal of the distance matrix in place of the
+    sliding window dot product found in the original STOMP method.
+    """
+    m_inverse = 1.0 / m
+    constant = (m - 1) * m_inverse * m_inverse  # (m - 1)/(m * m)
+    uint64_m = np.uint64(m)
+
+    # for storing diagonal iteration ranges
+    iter_ranges = np.zeros((config.STUMPY_N_DIAGONALS, 3), dtype=np.int64)
+    # for storing covariance values while iterating in chunks
+    covariances = np.zeros(config.STUMPY_N_DIAGONALS, dtype=np.float64)
+    # for storing range of diagonals to traverse in chunks
+    chunk_diags_range = np.arange(config.STUMPY_N_DIAGONALS)
+
+    for tile_idx in range(tiles_ranges[thread_idx, 0], tiles_ranges[thread_idx, 1]):
+        (
+            tile_i,  # i location of current tile in distance profile
+            tile_j,  # j location of current tile in distance profile
+            tile_height,  # height of current tile
+            tile_width,  # width of current tile
+            tile_start_diag,  # index of first diagonal to traverse on current tile
+            tile_stop_diag,  # index of last diagonal to traverse on current tile + 1
+            _,
+        ) = tiles[tile_idx]
+
+        # chunk_start_idx is the starting index of the chunk we are dealing with
+        for chunk_start_idx in range(
+            tile_start_diag,
+            tile_stop_diag,
+            config.STUMPY_N_DIAGONALS,
+        ):
+            # number of diagonals to traverse together
+            chunk_size = min(
+                config.STUMPY_N_DIAGONALS,
+                tile_stop_diag - chunk_start_idx,
+            )
+            # range of diags to traverse in chunk
+            chunk_diags = chunk_diags_range[:chunk_size]
+            # range of diags to traverse in chunk, with respect to entire distance
+            # matrix
+            tile_diags = chunk_start_idx + chunk_diags
+            # store the diagonal indices we are dealing with, with respect to entire
+            # distance matrix
+            iter_ranges[chunk_diags, 0] = tile_diags - tile_i + tile_j
+            # store the i locations where the diagonals begin
+            iter_ranges[chunk_diags, 1] = tile_i - np.minimum(0, tile_diags)
+            # store the i locations where the diagonals end
+            iter_ranges[chunk_diags, 2] = tile_i + np.minimum(
+                tile_height, tile_width - tile_diags
+            )
+            # record length of longest diagonal
+            longest_diag_len = np.max(
+                iter_ranges[chunk_diags, 2] - iter_ranges[chunk_diags, 1]
+            )
+
+            # row idx is the index of the row of n diagonals we are traversing together
+            for row_idx in range(longest_diag_len):
```

Comment:
If I understand correctly, technically EVERYTHING below this part should be nearly identical to `stump._compute_diagonals` except that we compute chunks of diagonals here while `stump._compute_diagonals` computes single diagonals. First, I want to split this part out as `stump_tiles._compute_diagonals`. Then, is there a way to cleanly combine `stump._compute_diagonals` and `stump_tiles._compute_diagonals` into a single function so that we aren't repeating code?

Reply:
@alvii147 
So, just to make sure that I am at the same page as you and @seanlaw are, first allow me to confirm my understanding:
* In existing `stumpy/stump.py`, we traverse one chunk of diags in one thread, and _**then we traverse that chunk diagonally!**_ 

* Here, in `stump_tiles`, we traverse a number of tiles in a thread. For each tile, we traverse chunks of its diagonals, and _**then we traverse each chunk  in a row-wise manner!**_


So, while, at some point in the algorithm, we have a chunk of diagonal in both `stump.py` and `stump_tiles.py`, the former traverse each chunk diagonally but the latter traverse each chunk row-by-row!

If my understanding is correct, then I think the proposal of adding function `update_profile` sounds good. We just wrap a function around that part of the code that deals with calculating new pearson value for the pair `(subseq i, subseq j)`

Reply:
> stump.compute_diagonals and stump_tiles.compute_tiles are functions that traverse things very differently

I see your point now. In `stump`, the load balancing happens at the diagonal level while, in `stump_tile`, it happens at the tile level.

Reply:
1. Let's keep `stump` and `stump_tiles` separate for now
2. Do split up `_compute_tiles` into two functions so that the actual distance computation part that looks like `stump._compute_diagonals` is in another function

Once it's broken out, I think it will be easier to read and reason with and may allow us to see how things might be potentially (or not) refactored. Does that work?

Reply:
> first allow me to confirm my understanding

@NimaSarajpoor your understanding is correct 🙃 

> I see your point now. In stump, the load balancing happens at the diagonal level while, in stump_tile, it happens at the tile level.

@seanlaw yepp that's right!

> Do split up _compute_tiles into two functions so that the actual distance computation part that looks like stump._compute_diagonals is in another function

That makes sense, I'll do that.